mailbox/api/src/main/java/org/apache/james/mailbox/model/ContentType.java

```diff
@@ -19,15 +19,168 @@
 
 package org.apache.james.mailbox.model;
 
+import java.io.StringReader;
+import java.nio.charset.Charset;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.apache.james.mime4j.dom.field.ContentTypeField;
+import org.apache.james.mime4j.field.Fields;
+import org.apache.james.mime4j.field.contenttype.parser.ContentTypeParser;
+import org.apache.james.mime4j.field.contenttype.parser.ParseException;
+
+import com.google.common.base.MoreObjects;
 import com.google.common.base.Preconditions;
 import com.google.common.base.Strings;
 
 public class ContentType {
+    public static class MimeType {
+        public static MimeType of(String value) {
+            ContentTypeParser parser = new ContentTypeParser(new StringReader(value));
+            try {
+                parser.parseAll();
+            } catch (ParseException e) {
+                throw new IllegalArgumentException("Invalid mimeType", e);
+            }
+            return new MimeType(
+                new MediaType(parser.getType()),
+                new SubType(parser.getSubType()));
+        }
+
+        public static MimeType of(MediaType mediaType, SubType subType) {
+            return new MimeType(mediaType, subType);
+        }
+
+        private final MediaType mediaType;
+        private final SubType subType;
+
+        private MimeType(MediaType mediaType, SubType subType) {
+            this.mediaType = mediaType;
+            this.subType = subType;
+        }
+
+        public String asString() {
+            return mediaType.asString() + "/" + subType.asString();
+        }
+
+        @Override
+        public final boolean equals(Object o) {
+            if (o instanceof MimeType) {
+                MimeType mimeType = (MimeType) o;
+
+                return Objects.equals(this.mediaType, mimeType.mediaType)
+                    && Objects.equals(this.subType, mimeType.subType);
+            }
+            return false;
+        }
+
+        @Override
+        public final int hashCode() {
+            return Objects.hash(mediaType, subType);
+        }
+
+        @Override
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                .add("mediaType", mediaType)
+                .add("subType", subType)
+                .toString();
+        }
+    }
+
+    public static class MediaType {
```

Comment:
FYI there are already some implementations handling all of that, such as guava MediaType. Would it be worth to use them?

Reply:
We are in mailbox api, we should refrain from coupling ourself to some external APIs ?

Reply:
We can IMO use it internally for parsing things, but the API should IMO not contain them.

Internally I made the choice of relying on mime4J

Reply:
IMO it's ok to publish types from mime4j as mime4j is a part of James

Reply:
mailbox-api is already relying on Guava...

Reply:
> mailbox-api is already relying on Guava...

It's using it to implement stuff but I bet we don't expose Guava classes as part of our interface.

> IMO it's ok to publish types from mime4j as mime4j is a part of James

That's unfortunate that mime4J have no strong types for MediaType, SubType, Charset.

(And IMO it starts being borderline)

Reply:
Let's change mime4j (: 
(as it's a longer process, I'm ok to start with current solution)

Reply:
> Let's change mime4j (:

That's the strategic move.

I propose to enact my work as a tactical move, and open a JIRA ticket for Mime4J breaking change API strong typing that might deserve a major release.

Reply:
See https://issues.apache.org/jira/browse/MIME4J-292